File: backend/app/api/v1/analytics.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.base import get_db
from app.schemas.analytics import (
    ActivityItem,
    AiSummaryUsageResponse,
    CallsByDayResponse,
    CategoryBreakdownResponse,
    EscalationRateResponse,
    KpiReport,
    KpiValue,
    PriorityBreakdownResponse,
    RecentActivityResponse,
    SourceBreakdownResponse,
)
from app.services import analytics_service

router = APIRouter(prefix="/analytics", tags=["analytics"])

# Matches WIREFRAMES.md's Analytics page default range.
DaysWindow = Query(30, ge=1, le=365)


@router.get("/kpis", response_model=KpiReport)
async def get_kpis(db: AsyncSession = Depends(get_db)) -> KpiReport:
    data = await analytics_service.get_kpis(db)
    return KpiReport(**{key: KpiValue(**value) for key, value in data.items()})


@router.get("/recent-activity", response_model=RecentActivityResponse)
async def get_recent_activity(
    limit: int = Query(10, ge=1, le=50),
    db: AsyncSession = Depends(get_db),
) -> RecentActivityResponse:
    tickets = await analytics_service.get_recent_activity(db, limit=limit)
    items = [
        ActivityItem(
            type="ticket_created",
            occurred_at=ticket.created_at,
            ticket_id=ticket.id,
            ticket_number=ticket.ticket_number,
            caller_name=ticket.caller_name,
            category=ticket.category.name,
            priority=ticket.priority,
            status=ticket.status,
        )
        for ticket in tickets
    ]
    return RecentActivityResponse(items=items)


@router.get("/tickets-by-category", response_model=CategoryBreakdownResponse)
async def tickets_by_category(days: int = DaysWindow, db: AsyncSession = Depends(get_db)) -> CategoryBreakdownResponse:
    items = await analytics_service.get_tickets_by_category(db, days=days)
    return CategoryBreakdownResponse(items=items)


@router.get("/tickets-by-priority", response_model=PriorityBreakdownResponse)
async def tickets_by_priority(days: int = DaysWindow, db: AsyncSession = Depends(get_db)) -> PriorityBreakdownResponse:
    items = await analytics_service.get_tickets_by_priority(db, days=days)
    return PriorityBreakdownResponse(items=items)


@router.get("/tickets-by-source", response_model=SourceBreakdownResponse)
async def tickets_by_source(days: int = DaysWindow, db: AsyncSession = Depends(get_db)) -> SourceBreakdownResponse:
    items = await analytics_service.get_tickets_by_source(db, days=days)
    return SourceBreakdownResponse(items=items)


@router.get("/calls-by-day", response_model=CallsByDayResponse)
async def calls_by_day(days: int = DaysWindow, db: AsyncSession = Depends(get_db)) -> CallsByDayResponse:
    items = await analytics_service.get_calls_by_day(db, days=days)
    return CallsByDayResponse(items=items)


@router.get("/escalation-rate", response_model=EscalationRateResponse)
async def escalation_rate(days: int = DaysWindow, db: AsyncSession = Depends(get_db)) -> EscalationRateResponse:
    data = await analytics_service.get_escalation_rate(db, days=days)
    return EscalationRateResponse(**data)


@router.get("/ai-summary-usage", response_model=AiSummaryUsageResponse)
async def ai_summary_usage(days: int = DaysWindow, db: AsyncSession = Depends(get_db)) -> AiSummaryUsageResponse:
    items = await analytics_service.get_ai_summary_usage(db, days=days)
    return AiSummaryUsageResponse(items=items)

File: frontend/src/api/analytics.ts
import { useQuery } from "@tanstack/react-query";
import type {
  AiSummaryUsageItem,
  CallsByDayItem,
  CategoryBreakdownItem,
  EscalationRateResponse,
  KpiReport,
  PriorityBreakdownItem,
  RecentActivityResponse,
  SourceBreakdownItem,
} from "../types/analytics";
import { ApiError } from "./client";

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL ?? "http://localhost:8000/api/v1";

async function get<T>(path: string): Promise<T> {
  const response = await fetch(`${API_BASE_URL}${path}`);
  if (!response.ok) {
    throw new ApiError(response.statusText, response.status);
  }
  return (await response.json()) as T;
}

export const analyticsApi = {
  getKpis: () => get<KpiReport>("/analytics/kpis"),
  getRecentActivity: (limit = 10) => get<RecentActivityResponse>(`/analytics/recent-activity?limit=${limit}`),
  getTicketsByCategory: (days = 30) => get<{ items: CategoryBreakdownItem[] }>(`/analytics/tickets-by-category?days=${days}`),
  getTicketsByPriority: (days = 30) => get<{ items: PriorityBreakdownItem[] }>(`/analytics/tickets-by-priority?days=${days}`),
  getTicketsBySource: (days = 30) => get<{ items: SourceBreakdownItem[] }>(`/analytics/tickets-by-source?days=${days}`),
  getCallsByDay: (days = 30) => get<{ items: CallsByDayItem[] }>(`/analytics/calls-by-day?days=${days}`),
  getEscalationRate: (days = 30) => get<EscalationRateResponse>(`/analytics/escalation-rate?days=${days}`),
  getAiSummaryUsage: (days = 30) => get<{ items: AiSummaryUsageItem[] }>(`/analytics/ai-summary-usage?days=${days}`),
};

export function useKpisQuery() {
  return useQuery({ queryKey: ["analytics", "kpis"], queryFn: analyticsApi.getKpis, refetchInterval: 30_000 });
}

export function useRecentActivityQuery(limit = 10) {
  return useQuery({
    queryKey: ["analytics", "recent-activity", limit],
    queryFn: () => analyticsApi.getRecentActivity(limit),
  });
}

export function useTicketsByCategoryQuery(days: number) {
  return useQuery({ queryKey: ["analytics", "tickets-by-category", days], queryFn: () => analyticsApi.getTicketsByCategory(days) });
}

export function useTicketsByPriorityQuery(days: number) {
  return useQuery({ queryKey: ["analytics", "tickets-by-priority", days], queryFn: () => analyticsApi.getTicketsByPriority(days) });
}

export function useTicketsBySourceQuery(days: number) {
  return useQuery({ queryKey: ["analytics", "tickets-by-source", days], queryFn: () => analyticsApi.getTicketsBySource(days) });
}

export function useCallsByDayQuery(days: number) {
  return useQuery({ queryKey: ["analytics", "calls-by-day", days], queryFn: () => analyticsApi.getCallsByDay(days) });
}

export function useEscalationRateQuery(days: number) {
  return useQuery({ queryKey: ["analytics", "escalation-rate", days], queryFn: () => analyticsApi.getEscalationRate(days) });
}

export function useAiSummaryUsageQuery(days: number) {
  return useQuery({ queryKey: ["analytics", "ai-summary-usage", days], queryFn: () => analyticsApi.getAiSummaryUsage(days) });
}
